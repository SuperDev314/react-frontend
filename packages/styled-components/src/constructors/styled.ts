import constructWithOptions from './constructWithOptions';
import StyledComponent from '../models/StyledComponent';
import domElements from '../utils/domElements';

import { Target } from '../types';

type StyledTemplateFunction = ReturnType<typeof constructWithOptions>;

type ElementShortcuts = {
  [key in keyof JSX.IntrinsicElements]?: StyledTemplateFunction;
};

interface Styled extends ElementShortcuts {
  (tag: Target): StyledTemplateFunction;
}

const styled: Styled = (tag: Target) => constructWithOptions(StyledComponent, tag);

// Shorthands for all valid HTML Elements
domElements.forEach(domElement => {
  styled[domElement] = styled(domElement);
});

export default styled;
