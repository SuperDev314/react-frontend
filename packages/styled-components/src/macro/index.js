// @flow

import { createMacro, MacroError } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';
import { addDefault, addNamed } from '@babel/helper-module-imports';

function styledComponentsMacro({ references, state, babel: { types: t }, config = {} }) {
  const allowedImports = [
    'default',
    'createGlobalStyle',
    'css',
    'isStyledComponent',
    'keyframes',
    'ServerStyleSheet',
    'StyleSheetConsumer',
    'StyleSheetContext',
    'StyleSheetManager',
    'ThemeConsumer',
    'ThemeContext',
    'ThemeProvider',
    'withTheme'
  ];

  const program = state.file.path;

  // FIRST STEP : replace `styled-components/macro` by `styled-components
  // references looks like this
  // { default: [path, path], css: [path], ... }
  let customImportName;
  Object.keys(references).forEach(refName => {
    if (!allowedImports.includes(refName)) {
      throw new MacroError(
        `Invalid import: ${refName}. You can only import ${allowedImports.join(
          ', '
        )} from 'styled-components/macro'.`
      );
    }

    // generate new identifier
    let id;
    if (refName === 'default') {
      id = addDefault(program, 'styled-components', { nameHint: 'styled' });
      customImportName = id;
    } else {
      id = addNamed(program, refName, 'styled-components', { nameHint: refName });
    }

    // update references with the new identifiers
    references[refName].forEach(referencePath => {
      // eslint-disable-next-line no-param-reassign
      referencePath.node.name = id.name;
    });
  });

  // SECOND STEP : apply babel-plugin-styled-components to the file
  const stateWithOpts = { ...state, opts: config, customImportName };
  program.traverse(babelPlugin({ types: t }).visitor, stateWithOpts);
}

export default createMacro(styledComponentsMacro, {
  configName: 'styledComponents'
});
