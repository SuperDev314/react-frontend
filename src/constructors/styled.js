// @flow
import constructWithOptions from './constructWithOptions'
import type { Target } from '../types'
import domElements from '../utils/domElements'

export default (styledComponent: Function) => {
  const styled = (tag: Target) => constructWithOptions(styledComponent, tag)

  // Shorthands for all valid HTML Elements
  domElements.forEach(domElement => {
    styled[domElement] = styled(domElement)
  })

  return styled
}
