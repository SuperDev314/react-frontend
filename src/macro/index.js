// @flow
import { createMacro, MacroError } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import * as styled from '..';

const allowedImports: Array<string> = Object.keys(styled).filter(helper => helper !== '__esModule');

function styledComponentsMacro({ references, state, babel: { types: t }, config = {} }) {
  const program = state.file.path;

  // FIRST STEP : replace `styled-components/macro` by `styled-components
  // references looks like this
  // { default: [path, path], css: [path], ... }
  let customImportName;
  Object.keys(references).forEach(refName => {
    if (!allowedImports.includes(refName)) {
      throw new MacroError(
        `Invalid import: ${refName}. You can only import ${allowedImports.join(
          ', '
        )} from 'styled-components/macro'.`
      );
    }

    // generate new identifier
    let id;
    if (refName === 'default') {
      id = addDefault(program, 'styled-components', { nameHint: 'styled' });
      customImportName = id;
    } else {
      id = addNamed(program, refName, 'styled-components', { nameHint: refName });
    }

    // update references with the new identifiers
    references[refName].forEach(referencePath => {
      // eslint-disable-next-line no-param-reassign
      referencePath.node.name = id.name;
    });
  });

  // SECOND STEP : apply babel-plugin-styled-components to the file
  const stateWithOpts = { ...state, opts: config, customImportName };
  program.traverse(babelPlugin({ types: t }).visitor, stateWithOpts);
}

const configName = 'styledComponents';

export default createMacro(styledComponentsMacro, { configName });
