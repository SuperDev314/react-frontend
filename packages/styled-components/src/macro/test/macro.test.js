import cosmiconfigMock from 'cosmiconfig';
import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

jest.mock('cosmiconfig', () => jest.fn(require.requireActual('cosmiconfig')));

const styledExampleCode = `
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const customStyledExampleCode = `
import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const cssExampleCode = `
import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`
`;

const keyframesExampleCode = `
import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`
`;

const createGlobalStyleExampleCode = `
import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`
`;

const ThemeProviderExampleCode = `
import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider, 
  { theme: { color: 'red' }}, 
  'hello'
)
`;

const extendsExampleCode = `
import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`
`;

const requireExampleCode = `
const styled = require('../../macro')

styled.div\`
  background: red;
\`
`;

const invalidExampleCode = `
import { UnknownImport } from '../../macro'
`;

pluginTester({
  title: 'macro',
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    'should work with styled': styledExampleCode,
    'should work with custom import name': customStyledExampleCode,
    'should work with { css }': cssExampleCode,
    'should work with { keyframes }': keyframesExampleCode,
    'should work with { createGlobalStyle }': createGlobalStyleExampleCode,
    'should work with { ThemeProvider }': ThemeProviderExampleCode,
    'should work when extending a component': extendsExampleCode,
    'should work with require() to import styled-components': requireExampleCode,
    'should throw error when importing { UnknownImport }': {
      code: invalidExampleCode,
      error: true,
      snapshot: false,
    },
    'should not add componentId with a config disabling ssr': {
      code: styledExampleCode,
      setup: () => {
        cosmiconfigMock.mockImplementationOnce(() => ({
          searchSync: () => ({
            config: {
              styledComponents: {
                ssr: false,
              },
            },
          }),
        }));
      },
    },
  },
});
