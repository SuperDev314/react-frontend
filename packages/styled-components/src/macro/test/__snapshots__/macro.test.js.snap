// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro should not add componentId with a config disabling ssr: should not add componentId with a config disabling ssr 1`] = `
"
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";

_styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:\\", \\";\\"], p => p.error ? 'red' : 'green');
"
`;

exports[`macro should work when extending a component: should work when extending a component 1`] = `
"
import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import React from 'react';

const Hello = () => React.createComponent(div, null, 'hello');

_styled(Hello).withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:red;\\"]);
"
`;

exports[`macro should work with { ThemeProvider }: should work with { ThemeProvider } 1`] = `
"
import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider,
  { theme: { color: 'red' }},
  'hello'
)

      ↓ ↓ ↓ ↓ ↓ ↓

import { ThemeProvider as _ThemeProvider } from \\"styled-components\\";
React.createComponent(_ThemeProvider, {
  theme: {
    color: 'red'
  }
}, 'hello');
"
`;

exports[`macro should work with { createGlobalStyle }: should work with { createGlobalStyle } 1`] = `
"
import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { createGlobalStyle as _createGlobalStyle } from \\"styled-components\\";
_createGlobalStyle\`
  background: red;
\`;
"
`;

exports[`macro should work with { css }: should work with { css } 1`] = `
"
import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { css as _css } from \\"styled-components\\";

_css([\\"color:\\", \\";\\"], props => props.whiteColor ? 'white' : 'black');
"
`;

exports[`macro should work with { keyframes }: should work with { keyframes } 1`] = `
"
import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { keyframes as _keyframes } from \\"styled-components\\";

_keyframes([\\"0%{opacity:0;}100%{opacity:1;}\\"]);
"
`;

exports[`macro should work with custom import name: should work with custom import name 1`] = `
"
import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import styled from 'another-package';

_styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:\\", \\";\\"], p => p.error ? 'red' : 'green');
"
`;

exports[`macro should work with require() to import styled-components: should work with require() to import styled-components 1`] = `
"
const styled = require('../../macro')

styled.div\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";

_styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:red;\\"]);
"
`;

exports[`macro should work with styled: should work with styled 1`] = `
"
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";

_styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:\\", \\";\\"], p => p.error ? 'red' : 'green');
"
`;
